eth/downloader/peer.go

```diff
@@ -409,7 +419,12 @@ func (ps *peerSet) BodyIdlePeers() ([]*peer, int) {
 	idle := func(p *peer) bool {
 		return atomic.LoadInt32(&p.blockIdle) == 0
 	}
-	return ps.idlePeers(62, 64, idle)
+	throughput := func(p *peer) float64 {
```

Comment:
4 times :) Alas, there's no way to avoid it. When idle peers is called, we don't have a particular peer we want to get the throughput out of, there's no way to specify which of the 4 throughput values to load up in the generic code (apart from maybe reflection, but that's not very clean). 

We would need somehow to be able to specify for the `idlePeers` function that we want for example the `bodyThroughput` values, and the `idlePeers` should be able to fetch that value for any peer. It simply cannot do that without a peer idependent accessor method. Now, it's arguable that we don't need to duplicate the locks themselves, but imho it's safer coding this way as we can use the accessor methods any way seen fit and don't need to worry about synchronisation. The opposite would mean that we cannot simply compare 2 throughput values, because that would require locking 2 peers, which could deadlock easily if same locks are obtained by someone else in a different order.